tests/pipeline/test_pipeline_delegator.py

```diff
@@ -0,0 +1,56 @@
+import time
+from unittest.mock import Mock, call
+
+from snuba.datasets.factory import get_dataset
+from snuba.datasets.plans.single_storage import SingleStorageQueryPlanBuilder
+from snuba.datasets.storages import StorageKey
+from snuba.datasets.storages.factory import get_storage
+from snuba.pipeline.pipeline_delegator import PipelineDelegator
+from snuba.pipeline.simple_pipeline import SimplePipelineBuilder
+from snuba.query.parser import parse_query
+from snuba.request import Request
+from snuba.request.request_settings import HTTPRequestSettings
+from snuba.web import QueryResult
+
+
+def test() -> None:
+    query_result = QueryResult({}, {"stats": {}, "sql": ""})
+    mock_query_runner = Mock(return_value=query_result)
+    mock_callback_func = Mock()
+    query_body = {
+        "selected_columns": ["type", "project_id"],
+    }
+
+    events = get_dataset("events")
+    query = parse_query(query_body, events)
+
+    events_pipeline = SimplePipelineBuilder(
+        query_plan_builder=SingleStorageQueryPlanBuilder(
+            storage=get_storage(StorageKey.EVENTS)
+        ),
+    )
+
+    errors_pipeline = SimplePipelineBuilder(
+        query_plan_builder=SingleStorageQueryPlanBuilder(
+            storage=get_storage(StorageKey.ERRORS)
+        ),
+    )
+
+    delegator = PipelineDelegator(
+        query_pipeline_builders={"events": events_pipeline, "errors": errors_pipeline},
+        selector_func=lambda query: ["events", "errors"],
+        callback_func=mock_callback_func,
+    )
+
+    delegator.build_pipeline(
+        Request("", query, HTTPRequestSettings(), {}, ""), mock_query_runner
+    ).execute()
+
+    # Allow threads to finish
+    time.sleep(0.01)
```

Comment:
Just shutdown the executor with a timeout (longer than this) instead (with the shutdown method: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown).  That will hold until either the timeout passes or all futures are done. 
Relying on arbitrary time delays to test multithreaded code is unreliable and makes the test flaky (if the time to sleep is too short) or long (if it is too long).

Reply:
This isn't ideal, but I want to avoid shutting down the executor. If we add another test here that uses this it will fail and we have no way to restart a previously shut down executor so we can run more queries on it.

Reply:
The issue is that, with that sleep this will certainly become flaky. That's a certainty. We have a few sleep in some tests but they are where there is no alternative. Like waiting on a Kafka driven pipeline.

Here we own the multithreading code, we can certainly find an alternative.
I would look at the problem this way: 
Now that you split the threaded delegator, let's ensure we test that onne properly in a multi-threaded environment. You can have a test dedicated to that class (you have more access to what the callables and callback do this way).  You could pass a Condition object to the callback (https://docs.python.org/3/library/threading.html#event-objects) and wait on that with a timeout in the main test thread. Then you can  even use a second condition to unblock the callback from the main thread after you received the result so you test that you get the first result independently of when the callback runs.

Then, about the end to end test, you can either use the same trick of adding a Condition in the callback you pass, or use a fake threaded function delegator to execute it.

Aside from that, when you integrate this in the query process you may want to pay attention to how many threads they start and, in case, you may want to be able to disable the multi threaded delegator for the snuba e2e tests unless we start adding assertions on the results of the secondary pipeline.

Reply:
I'm not sure I agree that this sleep becoming flaky is a certainty since we're not actually running any ClickHouse query here, just mocking the results. It will be doing even less when we port this test to the other class. A minimal sleep being passed only forces the main thread to release the GIL. We can certainly do the assertion in the callback or something else fancy to work around it, but I'm not really convinced this is particularly problematic.

Reply:
The problem is that we cannot know which kind of environments the test is supposed to run into and which kinds of constraints such environments will have in terms of resources, thus the issue with waiting a fixed amount of time. It will be fine most of the times, but it does not seem it is worth relying on that. 
Now I think there is something a lot more interesting to test anyway by actually unlocking the callback after the main thread received the results to ensure the delegator does not regress and always returns results as soon as the first execution is done. See above for how to do that. 
If you implement that test, automatically you will not need the sleep.

Reply:
Updated to remove the sleep